week-07-rails-continued/day-02/module-01/readme.md

```diff
@@ -1 +1,259 @@
+    <%= f.submit "Sign Up" %>
+  <% end %>
+
+  ```
+
+  **Stretch Challenge:** Can you do this in a modal in the `site#index`? Would you need your `users#new` route or action any more?
+
+14. In the `users#create` action, create a user, then log them in by creating a new session. After the user is successfully created and the session is set, redirect to the user show page.
+
+  ```ruby
+  #
+  # app/controllers/users_controller.rb
+  #
+  class UsersController < ApplicationController
+
+    ...
+
+    def create
+      user = User.create(user_params)
+      session[:user_id] = user.id
+
+      # redirect_to "/users/#{user.id}"
+      # the line above can be refactored to use rails route helpers:
+      redirect_to user_path(user)
+    end
+
+    ...
+
+  end
+  ```
+
+  **Stretch Challenge:** Create a condition that checks if the user was saved correctly.
+
+  **Hint:** first build the user in memory with `User.new(user_params)`, then check `if user.save` proceed as normal `else` redirect to `"/users/new"` again. Run `rails g scaffold user email password` in a SEPARATE SAMPLE PROJECT for an example.
+
+15. The last step is to display attributes for the currently logged in user on the user show page. To do this, update the `users#show` method in the controller find a user in the database using the current `session[:user_id]`.
+
+  ```ruby
+  #
+  # app/controllers/users_controller.rb
+  #
+  class UsersController < ApplicationController
+
+    ...
+
+    def show
+      @user = User.find(session[:user_id])
+    end
+
+    ...
+
+  end
+  ```
+
+16. Now, we want to display data for the currently logged-in user in the view:
+
+  ```erb
+  <!-- app/views/users/show.html.erb -->
+
+  Welcome, <%= @user.email %>!
+  ```
+
+1. Lastly, add a `current_user` method at the application level and then try to use it in your splash page.
+
+  ```ruby
+    #  app/controllers/application_controller.rb
+
+    class ApplicationController < ActionController::Base
+      protect_from_forgery with: :exception
+
+      def current_user
+        @current_user ||= session[:user_id] && User.find_by_id(session[:user_id])
+      end
+
+      helper_method :current_user  #makes the current_user method available to views
+    end
+  ```
+
+  ```erb
+  <!-- splash.html.erb -->
+  
+  Welcome <%= current_user %>
+  ```
+
+## Stretch Challenges
+
+1. Create a logout action and allow users to logout. What do you have to do to the `session[:user_id]`? Where do you `redirect_to` after they log out.
+1. Create a login page at `sessions#new`. Use the `@user.authenticate` method from `has_secure_password` to create a `sessions#create` action. (Hint, you'll need a `SessionsController`)
+
+1. Create another resource that the user has_many of called `posts`. Have the logged in user be able to create new posts that belong to them.
+
+1. Add <a href="http://guides.rubyonrails.org/active_record_validations.html" target="_blank">validations</a> to your `User` model.
```

Comment:
The rest of these stretch challenges are great, but I would add one on authorization and reorder them a little:

1. Authorization -- don't let logged in user sign up or log in again
2. `password_confirmation`
3. "Remember me" and "reset password"
4. Validations
5. Flash messages
6. `edit` and `update`

(The reason for the reordering is to move the stuff they're learning in the afternoon to the end.)